xen/drivers/passthrough/arm/ipmmu-vmsa.c

```diff
@@ -1179,6 +1193,16 @@ static void ipmmu_detach_device(struct iommu_domain *io_domain,
 	struct ipmmu_vmsa_archdata *archdata = to_archdata(dev);
 	unsigned int i;
 
+#ifdef CONFIG_HAS_COPROC
+	/*
+	 * Don't allow to even touch uTLBs when deassigning coproc device. Because it
+	 * may be running in other domain at the moment. Only coproc's scheduler
+	 * is allowed to decide when it is time to switch IOMMU context.
+	 */
+	if (dev_iommu(dev)->is_coproc)
```

Comment:
IOMMU in Xen was used to passthrough/protect device (which can be assigned to only one domain) so far. We add ability to handle such devices as a coproc (which can be assigned to several domains). For IOMMU to recognize these devices and "is_coproc" was introduced. If there are other users of IOMMU which require yet another IOMMU behavior, to be honest, I can't even imagine these users, we will add another flag or introduce variable describing a type of device. I don't now at the moment, we will see.